Use smart pointers and add null checks in WebInspectorClient.cpp

charliewolfe opened the pull request:
#### bd887680e39bbeee7851815ee7dc90c1d8dbe4b0
<pre>
Use smart pointers and add null checks in WebInspectorClient.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=270929">https://bugs.webkit.org/show_bug.cgi?id=270929</a>
<a href="https://rdar.apple.com/123523102">rdar://123523102</a>

Reviewed by Sihui Liu and David Kilzer.

Crash logs indicate `m_page` can be null in `WebInspectorClient::inspectedPageDestroyed()`.

* Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
(WebKit::WebInspectorClient::~WebInspectorClient):
(WebKit::WebInspectorClient::inspectedPageDestroyed):
(WebKit::WebInspectorClient::frontendCountChanged):
(WebKit::WebInspectorClient::openLocalFrontend):
(WebKit::WebInspectorClient::bringFrontendToFront):
(WebKit::WebInspectorClient::didResizeMainFrame):
(WebKit::WebInspectorClient::highlight):
(WebKit::WebInspectorClient::hideHighlight):
(WebKit::WebInspectorClient::showPaintRect):
(WebKit::WebInspectorClient::showInspectorIndication):
(WebKit::WebInspectorClient::hideInspectorIndication):
(WebKit::WebInspectorClient::didSetSearchingForNode):
(WebKit::WebInspectorClient::elementSelectionChanged):
(WebKit::WebInspectorClient::timelineRecordingChanged):
(WebKit::WebInspectorClient::setDeveloperPreferenceOverride):
(WebKit::WebInspectorClient::setEmulatedConditions):
(WebKit::WebInspectorClient::drawRect):

Canonical link: <a href="https://commits.webkit.org/276093@main">https://commits.webkit.org/276093@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5e7d45393af021c28fa587bd168b36d175889cad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43716 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22764 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46155 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46360 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39820 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26642 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20172 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36102 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44290 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19808 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37652 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17084 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17343 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1773 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39900 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47915 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18746 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15328 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42903 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20154 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41586 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20345 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5971 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19780 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
```diff
@@ -69,53 +69,68 @@ WebInspectorClient::WebInspectorClient(WebPage* page)
+    RefPtr page = m_page.get();
+    if (!page)
+        return;
+
```
To avoid behavior change, this should probably be moved to after `m_paintRectLayers.clear();`.

szewai reviewed (approved):
r=me

ddkilzer reviewed (approved):
r=me

I would make the change that @szewai suggested as well (to take the `RefPtr` later in the function.  You could even do something like this to further narrow the scope:

```
    if (m_paintRectOverlay) {
        RefPtr page = m_page.get();
        if (page && page->corePage())
            page->corePage()->pageOverlayController().uninstallPageOverlay(*m_paintRectOverlay, PageOverlay::FadeMode::Fade);
    }
```

charliewolfe committed:
Use smart pointers and add null checks in WebInspectorClient.cpp
https://bugs.webkit.org/show_bug.cgi?id=270929
rdar://123523102

Reviewed by Sihui Liu and David Kilzer.

Crash logs indicate `m_page` can be null in `WebInspectorClient::inspectedPageDestroyed()`.

* Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
(WebKit::WebInspectorClient::~WebInspectorClient):
(WebKit::WebInspectorClient::inspectedPageDestroyed):
(WebKit::WebInspectorClient::frontendCountChanged):
(WebKit::WebInspectorClient::openLocalFrontend):
(WebKit::WebInspectorClient::bringFrontendToFront):
(WebKit::WebInspectorClient::didResizeMainFrame):
(WebKit::WebInspectorClient::highlight):
(WebKit::WebInspectorClient::hideHighlight):
(WebKit::WebInspectorClient::showPaintRect):
(WebKit::WebInspectorClient::showInspectorIndication):
(WebKit::WebInspectorClient::hideInspectorIndication):
(WebKit::WebInspectorClient::didSetSearchingForNode):
(WebKit::WebInspectorClient::elementSelectionChanged):
(WebKit::WebInspectorClient::timelineRecordingChanged):
(WebKit::WebInspectorClient::setDeveloperPreferenceOverride):
(WebKit::WebInspectorClient::setEmulatedConditions):
(WebKit::WebInspectorClient::drawRect):

Canonical link: https://commits.webkit.org/276093@main
